File: engines/python/src/g2048/ai/agents.py
"""The common agent interface and the built-in agents (random, greedy, expectimax).

Every agent implements ``decide(observation) -> Decision``; the HTTP agent
protocol (spec/AGENT_PROTOCOL.md) is a direct serialisation of this contract.
"""

from __future__ import annotations

import time

from ..board import DIRECTION_NAMES, board_to_hex, board_to_matrix, valid_moves
from ..game import SPEC_VERSION
from ..rng import Rng
from .bitboard import EMPTY_COUNT, MASK16, from_board, moves4
from .expectimax import ExpectimaxSearch

_clock = time.perf_counter


class GameMeta:
    __slots__ = ("seed", "spec_version", "game_id")

    def __init__(self, seed: int, spec_version: int = SPEC_VERSION, game_id: str | None = None) -> None:
        self.seed = seed
        self.spec_version = spec_version
        self.game_id = game_id


class Observation:
    """What an agent sees: the exponent board (row-major), score, move count and game metadata."""

    __slots__ = ("board", "score", "move_count", "meta")

    def __init__(self, board, score: int = 0, move_count: int = 0, meta: GameMeta | None = None) -> None:
        self.board = board
        self.score = score
        self.move_count = move_count
        self.meta = meta if meta is not None else GameMeta(0)


class Decision:
    """A chosen direction (0-3) plus free-form metrics (camelCase keys, JSON-serialisable)."""

    __slots__ = ("move", "metrics")

    def __init__(self, move: int, metrics: dict) -> None:
        self.move = move
        self.metrics = metrics

    def __repr__(self) -> str:
        return f"Decision(move={DIRECTION_NAMES[self.move]!r}, metrics={self.metrics!r})"


class Agent:
    """Agent protocol. Subclasses set ``descriptor`` and implement ``decide``."""

    descriptor: dict = {}

    def reset(self, meta: GameMeta) -> None:  # noqa: B027 - optional hook
        """Called once per game before the first decision."""

    def decide(self, obs: Observation) -> Decision:
        raise NotImplementedError


def _descriptor(id_: str, name: str, config: dict | None = None) -> dict:
    d = {"id": id_, "name": name, "version": "1.0.0", "kind": "builtin", "language": "python"}
    if config is not None:
        d["config"] = config
    return d


class RandomAgent(Agent):
    """SPEC §10 reference random agent."""

    def __init__(self) -> None:
        self.descriptor = _descriptor("builtin/random", "Random")
        self.rng = Rng.from_seed(0)

    def reset(self, meta: GameMeta) -> None:
        self.rng = Rng.from_seed((meta.seed ^ 0xA5A5A5A5) & 0xFFFFFFFF)

    def decide(self, obs: Observation) -> Decision:
        t = _clock()
        valid = valid_moves(obs.board)
        move = valid[self.rng.below(len(valid))]
        return Decision(move, {"timeUs": int((_clock() - t) * 1e6 + 0.5), "deterministic": True})


class GreedyAgent(Agent):
    """One-ply greedy: maximise empty cells after the move, ties to the lower direction."""

    def __init__(self) -> None:
        self.descriptor = _descriptor("builtin/greedy", "Greedy")

    def decide(self, obs: Observation) -> Decision:
        t = _clock()
        b = from_board(obs.board)
        valid = valid_moves(obs.board)
        best = valid[0] if valid else 0
        best_score = -1
        ec = EMPTY_COUNT
        for d, m in enumerate(moves4(b)):
            if m == b:
                continue
            empty = ec[m & MASK16] + ec[(m >> 16) & MASK16] + ec[(m >> 32) & MASK16] + ec[m >> 48]
            if empty > best_score:
                best_score = empty
                best = d
        return Decision(best, {"timeUs": int((_clock() - t) * 1e6 + 0.5), "deterministic": True})


class ExpectimaxAgent(Agent):
    def __init__(self, config: dict | None = None, name: str = "Expectimax") -> None:
        self.search = ExpectimaxSearch(config)
        self.descriptor = _descriptor("builtin/expectimax", name, self.search.config.to_dict())
        self.last_result = None

    def decide(self, obs: Observation) -> Decision:
        b = from_board(obs.board)
        r = self.search.search(b)
        self.last_result = r
        move = r.move
        if move is None:
            valid = valid_moves(obs.board)
            move = valid[0] if valid else 0
        return Decision(
            move,
            {
                "depth": r.depth,
                "nodes": r.nodes,
                "ttHits": r.tt_hits,
                "ttSize": r.tt_size,
                "timeUs": r.time_us,
                "values": r.values,
                "heuristic": self.search.breakdown(b),
                "deterministic": r.deterministic,
                "completedDepths": r.completed_depths,
            },
        )


BUILTIN_AGENTS = ("random", "greedy", "expectimax")


def create_builtin_agent(agent_id: str, config: dict | None = None) -> Agent:
    if agent_id == "random":
        return RandomAgent()
    if agent_id == "greedy":
        return GreedyAgent()
    if agent_id == "expectimax":
        return ExpectimaxAgent(config or {})
    raise ValueError(f"unknown agent {agent_id!r} (expected one of {', '.join(BUILTIN_AGENTS)})")


def to_agent_request(obs: Observation) -> dict:
    """Wire format sent to remote agents (AGENT_PROTOCOL.md AgentRequest)."""
    req = {
        "seed": obs.meta.seed,
        "specVersion": obs.meta.spec_version,
        "board": board_to_matrix(obs.board),
        "boardHex": board_to_hex(obs.board),
        "score": obs.score,
        "moveCount": obs.move_count,
        "validMoves": [DIRECTION_NAMES[d] for d in valid_moves(obs.board)],
    }
    if obs.meta.game_id is not None:
        req = {"gameId": obs.meta.game_id, **req}
    return req

File: packages/engine/src/ai/agents.ts
/**
 * The common agent interface. Every agent — built-in, remote (HTTP), worker,
 * Python/Rust/Go runner or LLM-backed — is driven through this contract, and
 * the HTTP agent protocol (spec/openapi.yaml, `AgentRequest`/`AgentResponse`)
 * is a direct serialisation of it.
 */
import { type Board, type Direction, DIRECTION_NAMES, boardToHex, boardToMatrix, validMoves } from '../board.ts';
import { Rng } from '../rng.ts';
import { fromBoard, bbMove, outLo, outHi } from './bitboard.ts';
import { type ExpectimaxConfig, ExpectimaxSearch, type SearchResult } from './expectimax.ts';
import type { HeuristicBreakdown } from './heuristic.ts';

export interface GameMeta {
  gameId?: string;
  seed: number;
  specVersion: number;
}

export interface Observation {
  /** exponents, row-major (the canonical representation) */
  board: Board;
  score: number;
  moveCount: number;
  meta: GameMeta;
}

export interface DecisionMetrics {
  depth?: number;
  nodes?: number;
  ttHits?: number;
  ttSize?: number;
  timeUs: number;
  values?: (number | null)[];
  heuristic?: HeuristicBreakdown;
  deterministic?: boolean;
  completedDepths?: number[];
  note?: string;
}

export interface Decision {
  move: Direction;
  metrics: DecisionMetrics;
}

export type AgentKind = 'builtin' | 'remote' | 'worker' | 'llm' | 'human';

export interface AgentDescriptor {
  id: string;
  name: string;
  version: string;
  kind: AgentKind;
  language: string;
  config?: Record<string, unknown>;
}

export interface Agent {
  readonly descriptor: AgentDescriptor;
  /** Called once per game before the first decision. */
  reset?(meta: GameMeta): void;
  /** Return a valid move. Only called when at least one move is valid. */
  decide(obs: Observation): Decision | Promise<Decision>;
}

const clock: () => number = typeof performance !== 'undefined' ? () => performance.now() : () => Date.now();

/** SPEC §10 reference random agent. */
export class RandomAgent implements Agent {
  readonly descriptor: AgentDescriptor = { id: 'builtin/random', name: 'Random', version: '1.0.0', kind: 'builtin', language: 'typescript' };
  private rng = Rng.fromSeed(0);

  reset(meta: GameMeta): void {
    this.rng = Rng.fromSeed((meta.seed ^ 0xa5a5a5a5) >>> 0);
  }

  decide(obs: Observation): Decision {
    const t = clock();
    const valid = validMoves(obs.board);
    const move = valid[this.rng.below(valid.length)];
    return { move, metrics: { timeUs: Math.round((clock() - t) * 1000), deterministic: true } };
  }
}

/** One-ply greedy: maximise empty cells after the move, ties to the lower direction. */
export class GreedyAgent implements Agent {
  readonly descriptor: AgentDescriptor = { id: 'builtin/greedy', name: 'Greedy', version: '1.0.0', kind: 'builtin', language: 'typescript' };

  decide(obs: Observation): Decision {
    const t = clock();
    const [lo, hi] = fromBoard(obs.board);
    let best: Direction = validMoves(obs.board)[0];
    let bestScore = -1;
    for (let d = 0; d < 4; d++) {
      if (!bbMove(lo, hi, d)) continue;
      let empty = 0;
      for (let i = 0; i < 8; i++) {
        if (((outLo >>> (4 * i)) & 0xf) === 0) empty++;
        if (((outHi >>> (4 * i)) & 0xf) === 0) empty++;
      }
      if (empty > bestScore) {
        bestScore = empty;
        best = d as Direction;
      }
    }
    return { move: best, metrics: { timeUs: Math.round((clock() - t) * 1000), deterministic: true } };
  }
}

export class ExpectimaxAgent implements Agent {
  readonly descriptor: AgentDescriptor;
  readonly search: ExpectimaxSearch;
  lastResult: SearchResult | null = null;

  constructor(config: Partial<ExpectimaxConfig> = {}, name = 'Expectimax') {
    this.search = new ExpectimaxSearch(config);
    this.descriptor = {
      id: 'builtin/expectimax',
      name,
      version: '1.0.0',
      kind: 'builtin',
      language: 'typescript',
      config: { ...this.search.config },
    };
  }

  decide(obs: Observation): Decision {
    const [lo, hi] = fromBoard(obs.board);
    const r = this.search.search(lo, hi);
    this.lastResult = r;
    const move = (r.move ?? validMoves(obs.board)[0]) as Direction;
    return {
      move,
      metrics: {
        depth: r.depth,
        nodes: r.nodes,
        ttHits: r.ttHits,
        ttSize: r.ttSize,
        timeUs: r.timeUs,
        values: r.values,
        heuristic: this.search.breakdown(lo, hi),
        deterministic: r.deterministic,
        completedDepths: r.completedDepths,
      },
    };
  }
}

export type BuiltinAgentId = 'random' | 'greedy' | 'expectimax';

export function createBuiltinAgent(id: BuiltinAgentId, config: Record<string, unknown> = {}): Agent {
  switch (id) {
    case 'random':
      return new RandomAgent();
    case 'greedy':
      return new GreedyAgent();
    case 'expectimax':
      return new ExpectimaxAgent(config as Partial<ExpectimaxConfig>);
  }
}

/** Wire format sent to remote agents (see spec/openapi.yaml AgentRequest). */
export interface AgentRequest {
  gameId?: string;
  seed: number;
  specVersion: number;
  board: number[][];
  boardHex: string;
  score: number;
  moveCount: number;
  validMoves: string[];
}

export function toAgentRequest(obs: Observation): AgentRequest {
  return {
    gameId: obs.meta.gameId,
    seed: obs.meta.seed,
    specVersion: obs.meta.specVersion,
    board: boardToMatrix(obs.board),
    boardHex: boardToHex(obs.board),
    score: obs.score,
    moveCount: obs.moveCount,
    validMoves: validMoves(obs.board).map((d) => DIRECTION_NAMES[d]),
  };
}
